File: 5_kyu/Directions_Reduction.jl
module Dir
    export dir_reduc
    
    OPPOSITE = Dict( 'N' => 'S', 'S' => 'N', 'W' => 'E', 'E' => 'W' )

    function dir_reduc(arr::Array{String})::Array{String}
        workbench = String[]
        sizehint!( workbench, length(arr) )
        for next in arr
            if !isempty(workbench) && next[1]==OPPOSITE[workbench[end][1]]
                pop!(workbench)
            else
                push!(workbench, next)
            end
        end
        workbench
    end

end

File: 5_kyu/Directions_Reduction.rs
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
enum Direction {
    North,
    East,
    West,
    South,
}

fn opposide(dir: Direction) -> Direction {
    match dir {
        Direction::North => Direction::South,
        Direction::East => Direction::West,
        Direction::West => Direction::East,
        Direction::South => Direction::North,
    }
}

fn dir_reduc(arr: &[Direction]) -> Vec<Direction> {
    let mut accu = Vec::<Direction>::with_capacity(arr.len());
    for &current in arr {
        if !accu.is_empty() && accu[accu.len() - 1] == opposide(current) {
            accu.pop();
        } else {
            accu.push(current);
        }
    }
    accu
}
